test_handler.go

```diff
@@ -86,3 +100,37 @@ func ParseSHAParam(r *http.Request) (runSHA string, err error) {
 	}
 	return runSHA, err
 }
+
+// getLastCompleteRunSHA returns the SHA[0:10] for the most recent run that complete for all of the given browser names.
+func getLastCompleteRunSHA(ctx context.Context, browserNames []string) (sha string, err error) {
+	baseQuery := datastore.
+		NewQuery("TestRun").
+		Order("-CreatedAt").
+		Limit(100).
+		Project("Revision")
+
+	// Map is sha -> browser -> seen yet?  - this prevents over-counting dupes.
+	runSHAs := make(map[string]map[string]bool)
+	for _, browser := range browserNames {
+		it := baseQuery.Filter("BrowserName = ", browser).Run(ctx)
+		for {
+			var testRun TestRun
+			_, err := it.Next(&testRun)
+			if err == datastore.Done {
+				break
```

Comment:
If I'm reading this correctly, it breaks once it finds a match across all 4 browsers, correct? So in practice it won't be querying 400 entities, it will end up being closer to 10-20?

Reply:
No - I can (potentially) tweak it to have 4 iterators, and one-by-one them in parallel, but currently it fully iterates 3 browsers, and partially iterates the 4th (falling out a little early)